### ISSUE TITLE
[Discussion] What Metrics do we want?

### ISSUE THREAD

**Issue by justusschock** [PyTorchLightning/pytorch-lightning#1275]
As discussed in #973 , we will probably start by implementing metrics as standalones.

This issue aims to discuss on what metrics we need and how we can implement this in a package structure.

Suggestions welcome.

My initial thought was to have a metrics package with subpackages for each research area like vision, text, audio etc.

CC @srush @Borda @williamFalcon @Darktex

As a start: 
For vision I'd like to have the following:
- [ ] accuracy
- [ ] precision
- [ ] recall
- [ ] f1 score
- [ ] roc
- [ ] auc
- [ ] dice coefficient
- [ ] panoptic quality
- [ ] IOU
- [ ] SSIM

**Comment by Borda** [comment]
I like the structure...

**Comment by SkafteNicki** [comment]
Dividing only into research areas would mean duplication of some metrics, for example accuracy is used more or less within all fields. I think it would be better to mainly divide into a regression and classification subpackage depending on targets begin continues or discrete. Specific metrics (like BLEU in NLP) could be in research-specific subpackeges.

**Comment by justusschock** [comment]
I agree, and these metrics (like accuracy) would not fall in any of these but remain in the base package.

I don't want to divide them into regression and classification and also have subpackages for all the research areas as it may become non-trivial where to find the desired metric.
Another thing we could think of is not having subpackages at all, but just one metric package containing them all (just like `torch.nn`)

**Comment by Borda** [comment]
> Dividing only into research areas would mean duplication of some metrics, for example accuracy is used more or less within all fields. I think it would be better to mainly divide into a regression and classification subpackage depending on targets begin continues or discrete. Specific metrics (like BLEU in NLP) could be in research-specific subpackeges.

I would rather avoid deep metric structures, one level is enough... So we can have general purpose like accuracy and the domain specific =)

**Comment by seandatasci** [comment]
cv: panoptic quality and IOU
Augmentation: affinity and diversity

**Comment by haotongye** [comment]
Some metrics may even be dataset-specific, e.g., F1 score for SQuAD (there are some preprocessing and special rules involved). For these kind of less general metrics, I think there should be a base `Metric` class for people to inherited on and create their own.

For some reference, [this](https://github.com/haotongye/pytorch-project-example/blob/master/common/metrics.py) is how I implement mine, and [this](https://github.com/pytorch/ignite/blob/master/ignite/metrics/metric.py) is from PyTorch Ignite.

Also, should losses be considered as some type of metrics?

**Comment by justusschock** [comment]
@haotongye 
I would say, that we shouldn't include dataset specific metrics here.
But I agree, we should have a base class metric (probably just a `torch.nn.Module` with some extras). This will, however, be hard for the functional interface.

For now, I wouldn't include losses, as this would really broaden the scope. Maybe we can do this afterwards in a separate effort.

**Comment by justusschock** [comment]
@seandatasci Can you link a paper or reference implementation for the affinity and diversity part? AFAIK there are several ways to calculate these...

**Comment by williamFalcon** [comment]
some requested:
- confusion matrix
- f1
- AUC/ROC
- rouge
- bleu

**Comment by SkafteNicki** [comment]
metrics for continues output:
* mean squared error (MSE) / root mean squared error (RMSE)
* mean absolute error (MAE)
* root mean squared Logarithmic Error (RMSLE)
* maxerror
* cosinesimilarity

would be nice to have:
* R2 score (coefficient of determination) 
* Correlation (Pearson/Spearman)
* Explained variance score

however, as far as I know the last 3 require access to the full list of targets and predictions at ones, so they can only be used for smaller datasets.

**Comment by shubhamagarwal92** [comment]
- As I mentioned in the [tweet](https://twitter.com/shubhamag1992/status/1244595634642001921) for NLG [this repo](https://github.com/Maluuba/nlg-eval) can directly be integrated (?) 
 
- If planning to also include support for Vision&Language tasks such as VQA/Visdial etc. which are proposed mostly as discriminative tasks, R@{1,5,10} / MRR/ NDCG  can also be used. One nice implementation by Pythia [here](https://github.com/facebookresearch/pythia/blob/master/pythia/modules/metrics.py).

Let me know if I can help! Thanks.

**Comment by justusschock** [comment]
@shubhamagarwal92 We probably will have to adjust the metrics for NLG according to our upcoming metrics interface, but other then that it should be fine. If you want to, you can take this, once we have our interface running (probably tomorrow).

**Comment by seandatasci** [comment]
@justusschock https://arxiv.org/abs/2002.08973

**Comment by Borda** [comment]
> Let me know if I can help! Thanks.

Help is always welcome =)
